### ISSUE TITLE
"Letter of" doesn't work for all emoji/languages

### ISSUE THREAD

**Issue by paulkaplan** [LLK/scratch-vm#1418]
### Expected Behavior

_Please describe what should happen_

Not 100% sure, but I would expect that the "juggling clown" emoji would work the same way the "checkmark" one does when using "letter _ of _"

### Actual Behavior

_Describe what actually happens_

The juggling clown returns an "unknown character" fill-in, while the checkmark returns the checkmark correctly.

### Steps to Reproduce

_Explain what someone needs to do in order to see what's described in *Actual behavior* above_

![emoji-letter-of](https://user-images.githubusercontent.com/654102/43646543-323ef678-9703-11e8-9f31-84e129098691.gif)

### Operating System and Browser

_e.g. Mac OS 10.11.6 Safari 10.0_

Mac emoji picker. Probably has something to do with the "modifiers" (i.e. male/female clown options) , but I don't have a good understanding about how emoji work. 

This issue almost certainly effects other non-latin character sets as well, as @kchadha and @chrisgarrity brought up, making the "letter _ of _" block not useful in those languages.

For more about this, see docs on https://github.com/orling/grapheme-splitter

**Comment by Kenny2github** [comment]
In Python, the check mark emoji (✅) is represented with the Unicode escape `\u2705`.
The clown (actually "juggling") emoji (🤹) is represented with the *big* Unicode escape `\U0001f939`.
The check mark emoji is 3 bytes of Unicode (`%E2%9C%85`); the juggling emoji is 4 (`%F0%9F%A4%B9`). This leads me to believe that the block treats three bytes as a character and therefore thinks the juggling emoji is one three-byte character (`%F0%9F%A4`) and one one-byte character (`%B9`). Probable fix would be to make sure everything is properly encoded as Unicode? I'm not sure how the block works so I can't say for sure.

**Comment by thisandagain** [comment]
Yeah. This kind of stuff with emoji is always a little tricky. Luckily, there are a few NPM packages that have already done the legwork for handling this that we may want to consider:

https://www.npmjs.com/package/emoji-aware

**Comment by joker314** [comment]
In JavaScript, the juggling emoji is two characters (a "surrogate pair").

```
> "🤹".split``
< (2) ["�", "�"]
```

The two characters it's made from actually aren't two halves of the byte. Instead, they're determined through mathematical formulae. (There are entire Unicode blocks for high and low surrogate pairs).

Emojis can be longer than 2 characters, though, for example: modifiers (to determine skin tone, gender, direction, etc.) can be added to an emoji to change it. It's probably a good idea to use someone else's code, like thisandagain has suggested.

Note that

```
"🤹".length === 2 // true
```

So it's a good idea to investigate whether the `length of ( )` block is also affected (and I bet that it is)

**Comment by joker314** [comment]
Yup.

![Length of juggling emoji block results in 2](https://user-images.githubusercontent.com/18113170/44348410-feb91600-a4a2-11e8-99da-3cac6bdc6620.png)

**Comment by joker314** [comment]
Yup.

![Length of juggling emoji block results in 2](https://user-images.githubusercontent.com/18113170/44348410-feb91600-a4a2-11e8-99da-3cac6bdc6620.png)

**Comment by joker314** [comment]
Yup.

![Length of block returns 2 for juggling emoji](https://user-images.githubusercontent.com/18113170/44356672-aab92c00-a4b8-11e8-819a-de6fb9be9132.png)

**Comment by joker314** [comment]
A cool thing is Array.from() which utilises the feature that JavaScript's strings are iterators which handle surrogate pairs automatically.

```javascript
["I", " ", "💖", " ", "t", "o", " ", "🤹"] // Result
```

This then gives one the ability to find characters by index, as well as to compute length.

However, there are some weird things. The female astronaut emoji 👩‍🚀 can be created by combining the woman emoji 👩 and a rocket 🚀 using a zero-width joiner. So it's three characters (or in JS, five!). But `Array.from()` yields

```
["👩", "‍", "🚀"]
```

Hair, gender, and skin-tone modifiers will all be too confusing for JavaScript to handle. On the other hand, it's possible some projects might abuse this -- given that the same thing happened in 2.0.

Are there any projects that allow you to select the skin-tone of a man by doing

```
join (👨) (🏿)
```

and substituting the type 6 skin tone 🏿 with a variable?

I think a third-party library is best, to combine the female astronaut into one single character from a Scratcher's perspective?

**Comment by joker314** [comment]
("character" means "JavaScript string of length 1")

**summary**: almost all modern languages are not affected by this bug. Lots (hopefully most common) Chinese, Japanese, and Korean characters are not affected. Other Chinese, Japanese, and Korean characters _are_ affected.
 
> This issue almost certainly effects other non-latin character sets as well

Unicode is split into 17 planes. "Plane 0" (or BMP - Basic Multilingual Plane) contains most modern scripts. (The below lists some examples, non-exhaustively)
-  CJK characters: "20,976 basic Chinese characters" (for Chinese, Japanese, Korean)
- Latin script (extended): (for English, Spanish, Italian, German, etc.)
- Cyrillic script: (for Russian, Bulgarian, Ukrainian, Serbian, etc.)
- Armenian
- Aramaic scripts: (for Hebrew, Arabic, Samaritan [...], etc.)
- Tai scripts (for Tai Le, Buginese, etc.)
- _Lots more_

Sorry for inaccuracies in the above list.

**Plane 0 is not affected**. Every other plane is encoded in JavaScript via pairs of special characters in plane 0 -- the pair being called a "surrogate pair".

**Planes 1 and 2 are affected**. They mainly contain ancient scripts, or modern scripts for ancient languages, as well as emoji, musical symbols, and card symbols. However, the main exception to this is a lot of CJK characters. I don't know whether they are less common than those in plane 0, but that would be my assumption.

<hr>

As a quick fix, you could use the `Array.from` method I displayed in the above comment. This would allow for the CJK characters in planes 1 and 2.
